src/main/java/uk/gov/justice/digital/hmpps/keyworker/events/EventListener.java

```diff
@@ -1,53 +1,109 @@
-import java.util.Map;
+import java.util.Optional;
 
 @Service
 @ConditionalOnProperty(name = "sqs.provider")
 @Slf4j
+@AllArgsConstructor
 public class EventListener {
-
     private final ObjectMapper objectMapper;
     private final ReconciliationService reconciliationService;
-
-    public EventListener(final ReconciliationService reconciliationService, final ObjectMapper objectMapper) {
-        this.objectMapper = objectMapper;
-        this.reconciliationService = reconciliationService;
-    }
+    private final KeyworkerService keyworkerService;
 
     @JmsListener(destination = "${sqs.queue.name}")
-    public void eventListener(final String requestJson) {
-        final var event = getOffenderEvent(requestJson);
-        if (event != null) {
-            if ("EXTERNAL_MOVEMENT_RECORD-INSERTED".equals(event.getEventType())) {
+    public void eventListener(final String requestJson) throws JsonProcessingException {
+        final var message = getMessage(requestJson);
+        final var eventType = message.getMessageAttributes().getEventType().getValue();
+        log.info("Processing message of type {}", eventType);
+
+        final var event = getOffenderEvent(message.getMessage());
+        switch (eventType) {
+            case "EXTERNAL_MOVEMENT_RECORD-INSERTED":
                 reconciliationService.checkMovementAndDeallocate(event);
-            } else if ("BOOKING_NUMBER-CHANGED".equals(event.getEventType())) {
+                break;
+            case "BOOKING_NUMBER-CHANGED":
                 reconciliationService.checkForMergeAndDeallocate(event);
-            }
+                break;
+            case "DATA_COMPLIANCE_DELETE-OFFENDER":
+                Preconditions.checkState(StringUtils.isNotBlank(event.getOffenderIdDisplay()), String.format("Found null offender id for %s", requestJson));
+                keyworkerService.deleteKeyworkersForOffender(event.getOffenderIdDisplay());
+                break;
         }
     }
 
-    private OffenderEvent getOffenderEvent(final String requestJson) {
-        OffenderEvent event = null;
+    private Message getMessage(final String requestJson) throws JsonProcessingException {
         try {
-            final Map<String, String> message = objectMapper.readValue(requestJson, Map.class);
-            if (message != null && message.get("Message") != null) {
-                event = objectMapper.readValue(message.get("Message"), OffenderEvent.class);
-            }
-        } catch (IOException e) {
-            log.error("Failed to Parse Message {}", requestJson);
+            final var message = objectMapper.readValue(requestJson, new TypeReference<Message>() {
+            });
+            return Optional.ofNullable(message).orElseThrow(() -> new RuntimeException(String.format("Missing message from %s", requestJson)));
+        } catch (final JsonProcessingException e) {
+            log.error("Failed to parse Message {}", requestJson, e);
+            throw e;
```

Comment:
we now throw the exception when we get an error.  Previously we would swallow with a message.  This will mean that the message will get retried and put on the dead letter queue so that when we start monitoring the dead letter queue we can recognise the failures easier.

Reply:
will this lead to the exception being logged multiple times? here and in some global error handling?

Reply:
agreed, changed